### DISCUSSION TITLE
Is there any way to shutdown the runtime after terminating the running flowgraph?

### DISCUSSION THREAD

**Discussion by tkj666** [FutureSDR/FutureSDR#147] [Q&A] [answered]
After I use `FlowgraphHandler::terminate()` to stop the flowgraph, the blocks are all terminated but the runtime is still running, including control port and scheduler. Is there any way to shutdown the runtime gracefully?

**Comment by bastibl** [comment]
As long as the runtime and the scheduler exist, there are also the associated threads and services running. The runtime can spawn multiple flowgraphs, i.e., terminating the flowgraph should not stop it. If you drop the runtime (and with it its scheduler), it should shut down and terminate threads. For example, for the default scheduler: https://github.com/FutureSDR/FutureSDR/blob/main/src/runtime/scheduler/smol.rs#L40-L47

Let me know, if this design causes problems for your application.

**Reply by tkj666** [reply]
Thanks for the reply! I have tried to drop the runtime by leaving the scope and manually calling drop, but everything still seem to be running. All the threads are running, control port is still accessible and SmolSchedulerInner::drop is not called. Here is a simple snippet to reproduce.
```rust
fn main() {
    {
        let rt = Runtime::new();
        drop(rt);
    }
    println!("out");
    loop {}
}
```
It simply doesn't look like the intended behaviour...

**Reply by bastibl** [reply, marked as answer]
Arg, you're right. The `Runtime` is dropped, but `ControlPort` was not terminated. And since `ControlPort` also holds a reference to the Scheduler, the threads stick around. The most recent commit should implement correct shutdown of `ControlPort`. Maybe you could give it a try.
Thanks for bringing this up.

**Reply by tkj666** [reply]
Is it pushed to github? The latest commit is from 2 weeks ago for license.

**Reply by bastibl** [reply]
It was only on my fork, sorry. But now it should be there xD
